### PR TITLE
Monthly averaging callback

### PR DESCRIPTION
**Pull request by LenkaNovak**
# PULL REQUEST

## Purpose and Content
Adds ability to time-average fields using `@calendar_callback`.  This is a temporary solution for AMIP output until [ClimaDiagnostics](https://github.com/CliMA/ClimaDiagnostics/issues/1)  is developed. 

## Benefits and Risks
(State concisely the benefits to be derived from and the risks associated with this PR)
- b: addresses a required point of the AMIP roadmap
- r: resources spent in a temporary solution (until diagnostics across all models are unified) 

## Linked Issues
(Provide references to any link issues. Use closes #issuenum to automatically close an open issue)
- part of #74 

## PR Checklist
- [x] This PR has a corresponding issue OR is linked to an SDI.
- [x] I have followed CliMA's codebase [contribution](https://clima.github.io/ClimateMachine.jl/latest/Contributing/) and [style](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) guidelines OR N/A.
- [x] I have followed CliMA's [documentation policy](https://github.com/CliMA/policies/wiki/Documentation-Policy).
- [x] I have checked all issues and PRs and I certify that this PR does not duplicate an open PR.
- [x] I linted my code on my local machine prior to submission OR N/A.
- [x] Unit tests are included OR N/A.
- [x] Code used in an integration test OR N/A.
- [x] All tests ran successfully on my local machine OR N/A.
- [x] All classes, modules, and function contain docstrings OR N/A.
- [x] Documentation has been added/updated OR N/A.

### CODE REVIEW

File: experiments/AMIP/moist_mpi_earth/Manifest.toml
Comment (charleskawczynski) [inline experiments/AMIP/moist_mpi_earth/Manifest.toml:204]: Can we use the released version? Why are we cherry-picking a commit here?

File: experiments/AMIP/moist_mpi_earth/coupler_driver.jl
Comment (charleskawczynski) [inline experiments/AMIP/moist_mpi_earth/coupler_driver.jl:48]:
```suggestion
COUPLER_OUTPUT_DIR = joinpath(pkg_dir, "experiments/AMIP/moist_mpi_earth/output", mode_name)
mkpath(COUPLER_OUTPUT_DIR)
```
I think this should work, too? This is probably simpler

File: experiments/AMIP/moist_mpi_earth/coupler_driver.jl
Comment (charleskawczynski) [inline experiments/AMIP/moist_mpi_earth/coupler_driver.jl:216]: Macro hygiene is really hard IMO, I'd prefer we avoid the macro here

File: experiments/AMIP/moist_mpi_earth/coupler_driver.jl
Comment (charleskawczynski) [inline experiments/AMIP/moist_mpi_earth/coupler_driver.jl:265]:
```suggestion
    mkpath(COUPLER_OUTPUT_DIR*"_artifacts")
```

File: experiments/AMIP/moist_mpi_earth/coupler_driver.jl
Comment (charleskawczynski) [inline experiments/AMIP/moist_mpi_earth/coupler_driver.jl:277]: Let's use `mkdir` or `mkpath` instead of `run` for all of these

File: experiments/AMIP/moist_mpi_earth/Manifest.toml
Comment (LenkaNovak) [reply experiments/AMIP/moist_mpi_earth/Manifest.toml:204]: This is part of #118 There were some dependency issues in another PR, so we decided to hold off with the updates for the next two PRs (#118 should be mergable in the next couple of days :) ).

File: experiments/AMIP/moist_mpi_earth/coupler_driver.jl
Comment (LenkaNovak) [reply experiments/AMIP/moist_mpi_earth/coupler_driver.jl:48]: replaced with `isdir(COUPLER_OUTPUT_DIR*"_artifacts") || mkdir(COUPLER_OUTPUT_DIR*"_artifacts")`

File: experiments/AMIP/moist_mpi_earth/coupler_driver.jl
Comment (LenkaNovak) [reply experiments/AMIP/moist_mpi_earth/coupler_driver.jl:216]: We will address this as part of the clean up effort. This method was chosen because `@btime` was much faster than the function option (which was needed for operations at every timestep), but we did realise that this poses= some restrictions on the design, from the macro hygiene POV. So far our tests should catch problems with the macro hygiene but it is on the [agenda](https://github.com/CliMA/ClimaCoupler.jl/issues/92) to review this, along with fresh performance tests.

Comment (charleskawczynski) [reply experiments/AMIP/moist_mpi_earth/coupler_driver.jl:216]: `@btime` of what was much faster? and what was the wall clock time? Does a non-macro version exist?

File: experiments/AMIP/moist_mpi_earth/Manifest.toml
Comment (charleskawczynski) [reply experiments/AMIP/moist_mpi_earth/Manifest.toml:204]: What dependency issues?

Comment (LenkaNovak) [reply experiments/AMIP/moist_mpi_earth/Manifest.toml:204]: There was a behavior change for the updated packages, so we wanted to address this after merging in the regridding PR, this PR and the snow PR, as a separate issue.

File: experiments/AMIP/moist_mpi_earth/coupler_driver.jl
Comment (LenkaNovak) [reply experiments/AMIP/moist_mpi_earth/coupler_driver.jl:216]: `@btime` of a unit test of the @calendar_callback. I don't have the function on hand, because this was done months ago and okayed by reviewers. Since this is not part of this PR, can we address this separately?

Comment (charleskawczynski) [reply experiments/AMIP/moist_mpi_earth/coupler_driver.jl:216]: Yes, that's fine

File: experiments/AMIP/moist_mpi_earth/Manifest.toml
Comment (LenkaNovak) [reply experiments/AMIP/moist_mpi_earth/Manifest.toml]: Updates to Julia 1.8 and the current releases of CA, CC and SF done! :)

**Comment by LenkaNovak** [comment]
All above resolved and agreed to merge offline.

**Review by charleskawczynski — APPROVED** [review]

**Pull request merged**
